src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -1,171 +1,75 @@
-
-    // Block Offsets 
-    // (1 blockInRow = texelBlockDimensions.x texels)
-    // (1 rowInSlice = texelBlockDimensions.y texel rows)
-    // (1 sliceInLayer = texelBlockDimensions.z texel depths)
-    uint32_t currentBlockInRow = 0u;
-    uint32_t currentRowInSlice = 0u;
-    uint32_t currentSliceInLayer = 0u;
-    uint32_t currentLayerInRegion = 0u;
-    uint32_t currentRegion = 0u;
-            
-    // bufferOffsetAlignment:
-        // [x] If Depth/Stencil -> must be multiple of 4
-        // [ ] If multi-planar -> bufferOffset must be a multiple of the element size of the compatible format for the aspectMask of imagesubresource
-        // [x] If Queue doesn't support GRAPHICS_BIT or COMPUTE_BIT ->  must be multiple of 4
-        // [x] bufferOffset must be a multiple of texel block size in bytes
-    uint32_t bufferOffsetAlignment = texelBlockInfo.getBlockByteSize();
-    if(asset::isDepthOrStencilFormat(dstImage->getCreationParameters().format))
-        bufferOffsetAlignment = std::lcm(bufferOffsetAlignment, 4u);
-
-    bool queueSupportsCompute = (queueFamProps.queueFlags & IPhysicalDevice::EQF_COMPUTE_BIT).value != 0;
-    bool queueSupportsGraphics = (queueFamProps.queueFlags & IPhysicalDevice::EQF_GRAPHICS_BIT).value != 0;
-    if((queueSupportsGraphics || queueSupportsCompute) == false)
-        bufferOffsetAlignment = std::lcm(bufferOffsetAlignment, 4u);
-
-    // TODO: Need to have a function to get equivalent format of the specific plane of this format (in aspectMask)
-    // if(asset::isPlanarFormat(dstImage->getCreationParameters().format))
+    ImageRegionIterator regionIterator = ImageRegionIterator(regions, queueFamProps, srcBuffer, srcFormat, dstImage);
 
-    assert(core::is_alignment(bufferOffsetAlignment));
-    
+    // TODO[FUTURE]: consider benefits of using `limits->optimalBufferCopyOffsetAlignment` and `limits->optimalBufferCopyRowPitchAlignment`
     // Assuming each thread can handle minImageTranferGranularitySize of texelBlocks:
     const uint32_t maxResidentImageTransferSize = limits.maxResidentInvocations * texelBlockInfo.getBlockByteSize() * (minImageTransferGranularity.width * minImageTransferGranularity.height * minImageTransferGranularity.depth); 
-    // memoryLowerBound = max(maxResidentImageTransferSize, the largest rowPitch of regions); 
-    uint32_t memoryLowerBound = maxResidentImageTransferSize;
 
-    while (currentRegion < regions.size())
+    while (!regionIterator.isFinished())
     {
-        size_t memoryNeededForRemainingRegions = 0ull;
-        for (uint32_t i = currentRegion; i < regions.size(); ++i)
-        {
-            memoryNeededForRemainingRegions = core::alignUp(memoryNeededForRemainingRegions, bufferOffsetAlignment);
+        size_t memoryNeededForRemainingRegions = regionIterator.getMemoryNeededForRemainingRegions();
 
+        // memoryLowerBound = max(maxResidentImageTransferSize, the largest rowPitch of regions); 
```

Comment:
what's with the comment?

Reply:
probably my draft and thoughts, will get rid of it